### ISSUE TITLE
Pyright cannot handle imports in try-except in dask code

### ISSUE THREAD

**Issue by myuanz** [microsoft/pyright#6990]
In this code: 

```python
# pip install "dask[complete]"
# test.py
import dask.array as da

da.from_array
```

I got:
```bash
> pyright test.py
/home/myuan/projects/cell-label/test/test.py
  /home/myuan/projects/cell-label/test/test.py:5:4 - error: "from_array" is not exported from module "dask.array" (reportPrivateImportUsage)
1 error, 0 warnings, 0 informations 
```

And when I write `da.`, can only see `e` and `msg`:

![image](https://github.com/microsoft/pyright/assets/25359546/d01dc90b-da5f-4916-b90c-c526c9dc7945)

I checked the source code, the function I want to import does exists. It's just wrapped in a try-except block. https://github.com/dask/dask/blob/91dd42529b9ecd7139926ebadbf56a8f6150991f/dask/array/__init__.py#L16

---
pyright: 1.1.339
vscode: 1.85.1
pylance: v2023.12.1

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

The problem is not that it is imported in a try-except block. The problem is that `from_array` is imported by the `dask.array` module, but it's not re-exported from that module. The `dask` package is marked as a "py.typed" package, which means that it is assumed to follow the rules for typed libraries. If it had intended to re-export `from_array` from the `dask.array` module, it should have used a redundant form of import (i.e. `dask.array.core import from_array as from_array`). For details, refer to [this documentation](https://microsoft.github.io/pyright/#/typed-libraries?id=library-interface). 

If you change your code to `import dask.array.core as da`, it will work as expected.

**Comment by myuanz** [comment]
Thank you. I plan to submit some PRs to dask next.
